File: shared/src/commonMain/kotlin/com/ohmz/tday/shared/guide/GuideSearch.kt
package com.ohmz.tday.shared.guide

/**
 * The one guide search algorithm, shared by every platform.
 *
 * Matching mirrors the app's existing lowercase-substring search culture but adds
 * light weighting. To guarantee cross-platform parity, normalization is defined
 * here once: Android calls [buildDoc] at runtime, and the exporter calls the same
 * [normalize] to precompute the blobs shipped to web and iOS, so all three search
 * over byte-identical normalized text. The web (TS) and iOS (Swift) query-side
 * ports of [normalize] are kept honest by generated fixtures.
 */
object GuideSearch {

    /** A topic reduced to normalized, searchable text. */
    data class Doc(
        val topicId: String,
        val titleNorm: String,
        val keywordsNorm: String,
        val bodyNorm: String,
    )

    private const val TITLE_WEIGHT = 3
    private const val KEYWORD_WEIGHT = 2
    private const val BODY_WEIGHT = 1

    // Common Latin diacritics folded so "café"/"cafe" and "répéter"/"repeter" match.
    // Cyrillic and CJK need no folding — lowercase + raw substring already works.
    private val DIACRITIC_FOLD: Map<Char, Char> = buildMap {
        "àáâãäåāăą".forEach { put(it, 'a') }
        "çćčĉ".forEach { put(it, 'c') }
        "ďđ".forEach { put(it, 'd') }
        "èéêëēĕėęě".forEach { put(it, 'e') }
        "ìíîïĩīĭįı".forEach { put(it, 'i') }
        "ñńņň".forEach { put(it, 'n') }
        "òóôõöøōŏő".forEach { put(it, 'o') }
        "ùúûüũūŭůűų".forEach { put(it, 'u') }
        "ýÿŷ".forEach { put(it, 'y') }
        "śšşŝ".forEach { put(it, 's') }
        "žźż".forEach { put(it, 'z') }
        "ğĝ".forEach { put(it, 'g') }
        "ß".forEach { put(it, 's') }
    }

    /**
     * Lowercase, fold Latin diacritics, and collapse all whitespace runs to a
     * single space. Deterministic and dependency-free so every platform can
     * reproduce it exactly.
     */
    fun normalize(input: String): String {
        val lowered = input.lowercase()
        val sb = StringBuilder(lowered.length)
        var pendingSpace = false
        for (ch in lowered) {
            if (ch.isWhitespace()) {
                pendingSpace = sb.isNotEmpty()
                continue
            }
            if (pendingSpace) {
                sb.append(' ')
                pendingSpace = false
            }
            sb.append(DIACRITIC_FOLD[ch] ?: ch)
        }
        return sb.toString()
    }

    /** Build a normalized [Doc] from a topic's resolved localized strings. */
    fun buildDoc(topicId: String, title: String, keywords: String, body: String): Doc =
        Doc(
            topicId = topicId,
            titleNorm = normalize(title),
            keywordsNorm = normalize(keywords),
            bodyNorm = normalize(body),
        )

    /**
     * Rank [docs] against [query], returning matching topic ids best-first.
     *
     * A doc matches only if EVERY query token appears somewhere in it (token-AND).
     * Score sums, per token, a title hit (3) + keyword hit (2) + body hit (1).
     * Ties keep the input order, so pass [docs] in catalog order for a stable,
     * section-ordered tie-break. An empty query returns an empty list (callers
     * show the sectioned catalog instead).
     */
    fun rank(query: String, docs: List<Doc>): List<String> {
        val tokens = normalize(query).split(' ').filter { it.isNotEmpty() }
        if (tokens.isEmpty()) return emptyList()

        data class Scored(val topicId: String, val score: Int, val index: Int)

        val scored = ArrayList<Scored>(docs.size)
        for ((index, doc) in docs.withIndex()) {
            var total = 0
            var allTokensMatched = true
            for (token in tokens) {
                var tokenScore = 0
                if (doc.titleNorm.contains(token)) tokenScore += TITLE_WEIGHT
                if (doc.keywordsNorm.contains(token)) tokenScore += KEYWORD_WEIGHT
                if (doc.bodyNorm.contains(token)) tokenScore += BODY_WEIGHT
                if (tokenScore == 0) {
                    allTokensMatched = false
                    break
                }
                total += tokenScore
            }
            if (allTokensMatched) scored.add(Scored(doc.topicId, total, index))
        }
        // Higher score first; ties fall back to catalog (input) order.
        return scored
            .sortedWith(compareByDescending<Scored> { it.score }.thenBy { it.index })
            .map { it.topicId }
    }
}

File: tday-web/src/features/guide/guideSearch.ts
// TypeScript port of the shared Kotlin GuideSearch (shared/.../guide/GuideSearch.kt).
// It MUST stay behaviourally identical to the Kotlin engine — the generated
// fixtures in tests/fixtures/guide-search-vectors.json are produced by the Kotlin
// side and replayed here in tests/unit/guide-search.test.ts to catch any drift.

export interface GuideSearchDoc {
  topicId: string;
  titleNorm: string;
  keywordsNorm: string;
  bodyNorm: string;
}

const TITLE_WEIGHT = 3;
const KEYWORD_WEIGHT = 2;
const BODY_WEIGHT = 1;

// Same Latin diacritic groups as the Kotlin DIACRITIC_FOLD.
const FOLD_GROUPS: Array<[string, string]> = [
  ["àáâãäåāăą", "a"],
  ["çćčĉ", "c"],
  ["ďđ", "d"],
  ["èéêëēĕėęě", "e"],
  ["ìíîïĩīĭįı", "i"],
  ["ñńņň", "n"],
  ["òóôõöøōŏő", "o"],
  ["ùúûüũūŭůűų", "u"],
  ["ýÿŷ", "y"],
  ["śšşŝ", "s"],
  ["žźż", "z"],
  ["ğĝ", "g"],
  ["ß", "s"],
];

const DIACRITIC_FOLD: Map<string, string> = (() => {
  const map = new Map<string, string>();
  for (const [chars, base] of FOLD_GROUPS) {
    for (const ch of chars) map.set(ch, base);
  }
  return map;
})();

/** Lowercase, fold Latin diacritics, collapse whitespace runs to one space. */
export function normalize(input: string): string {
  const lowered = input.toLowerCase();
  let out = "";
  let pendingSpace = false;
  for (const ch of lowered) {
    if (/\s/.test(ch)) {
      pendingSpace = out.length > 0;
      continue;
    }
    if (pendingSpace) {
      out += " ";
      pendingSpace = false;
    }
    out += DIACRITIC_FOLD.get(ch) ?? ch;
  }
  return out;
}

export function buildDoc(
  topicId: string,
  title: string,
  keywords: string,
  body: string,
): GuideSearchDoc {
  return {
    topicId,
    titleNorm: normalize(title),
    keywordsNorm: normalize(keywords),
    bodyNorm: normalize(body),
  };
}

/**
 * Rank docs against a query, best-first. Every query token must appear somewhere
 * in a doc (token-AND); score sums per-token title(3)/keyword(2)/body(1) hits;
 * ties fall back to input order. Empty query returns []. Returns topic ids.
 */
export function rank(query: string, docs: GuideSearchDoc[]): string[] {
  const tokens = normalize(query)
    .split(" ")
    .filter((t) => t.length > 0);
  if (tokens.length === 0) return [];

  const scored: Array<{ topicId: string; score: number; index: number }> = [];
  docs.forEach((doc, index) => {
    let total = 0;
    let allMatched = true;
    for (const token of tokens) {
      let tokenScore = 0;
      if (doc.titleNorm.includes(token)) tokenScore += TITLE_WEIGHT;
      if (doc.keywordsNorm.includes(token)) tokenScore += KEYWORD_WEIGHT;
      if (doc.bodyNorm.includes(token)) tokenScore += BODY_WEIGHT;
      if (tokenScore === 0) {
        allMatched = false;
        break;
      }
      total += tokenScore;
    }
    if (allMatched) scored.push({ topicId: doc.topicId, score: total, index });
  });

  scored.sort((a, b) => b.score - a.score || a.index - b.index);
  return scored.map((s) => s.topicId);
}
